File: src/main/java/com/johnbryce/couponSystem/beans/Coupon.java
package com.johnbryce.couponSystem.beans;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.experimental.SuperBuilder;
import org.hibernate.Hibernate;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.Objects;

@Entity
@Table(name = "coupons")
@Data
@AllArgsConstructor
@SuperBuilder
public class Coupon extends Base {

    public Coupon() {
        super();
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @ManyToOne
    @JoinColumn(name="company_id")
    private Company company;

    @Enumerated(EnumType.STRING)
    private Category category;

    @Column(nullable = false, unique = true, length = 45)
    private String title;

    private String description;
    private Timestamp startDate;
    private Timestamp endDate;

    private int amount;

    private double price;
    private String image;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || Hibernate.getClass(this) != Hibernate.getClass(o)) return false;
        Coupon coupon = (Coupon) o;
        return id != 0 && Objects.equals(id, coupon.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: frontend/src/Models/Coupon.ts
import { number } from "yup/lib/locale";

export class Coupon{
    public id?:number;
    public companyId?:number;
    public title?:string;
    public description?:string;
    public category?:string;
    public startDate?:Date;
    public endDate?:Date;
    public amount?:number;
    public price?:number;
    public image?:string;

    public constructor(id?:number, companyId?:number, title?:string, description?:string, category?:string, startDate?:Date, endDate?:Date, amount?:number, price?:number, image?:string){
        this.id = id;
        this.companyId = companyId;
        this.title = title;
        this.description = description;
        this.category = category;
        this.startDate = startDate;
        this.endDate = endDate;
        this.amount = amount;
        this.price = price;
        this.image = image;
    }
}